### PR TITLE
[Remote Store] Fix tests when we restore index without any refresh

### PR DESCRIPTION
**Pull request by sachinpkale**
### Description
- If we have an index where all the ingested data is in translog alone (the first refresh is not triggered yet), restoring such index from remote store fails due to issue mentioned in https://github.com/opensearch-project/OpenSearch/issues/7923
- This happens as the translog UUID of the downloaded translog files do not match with that of segment files (that are created newly as part of restore)
- In this change, we make sure to create empty store with the same UUID that is present in the remote translog.

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/7923

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/store/Store.java [hunk 1]
```diff
@@ -1747,13 +1747,13 @@ public void accept(ShardLock Lock) {}
...
 
-    /**
-     * creates an empty lucene index and a corresponding empty translog. Any existing data will be deleted.
-     */
-    public void createEmpty(Version luceneVersion) throws IOException {
+    public void createEmpty(Version luceneVersion, String translogUUID) throws IOException {
         metadataLock.writeLock().lock();
         try (IndexWriter writer = newEmptyIndexWriter(directory, luceneVersion)) {
             final Map<String, String> map = new HashMap<>();
+            if (translogUUID != null) {
```
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/index/store/Store.java:1754]: I will add UTs around this change.

File: server/src/main/java/org/opensearch/index/translog/TranslogHeader.java [hunk 2]
```diff
@@ -137,9 +137,26 @@ static int readHeaderVersion(final Path path, final FileChannel channel, final S
     }
 
     /**
-     * Read a translog header from the given path and file channel
+     * Read a translog header from the given path and file channel and compare the given UUID
      */
     static TranslogHeader read(final String translogUUID, final Path path, final FileChannel channel) throws IOException {
+        TranslogHeader translogHeader = read(path, channel);
```
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/index/translog/TranslogHeader.java:143]: I will add UTs for this change.

File: server/src/main/java/org/opensearch/index/store/Store.java [hunk 3]
```diff
@@ -1747,13 +1747,13 @@ public void accept(ShardLock Lock) {}
         };
     }
 
-    /**
-     * creates an empty lucene index and a corresponding empty translog. Any existing data will be deleted.
-     */
-    public void createEmpty(Version luceneVersion) throws IOException {
+    public void createEmpty(Version luceneVersion, String translogUUID) throws IOException {
```
[the remark this answers is not in the document]
Comment (sachinpkale) [reply server/src/main/java/org/opensearch/index/store/Store.java:1750]: Yeah, needs to be public as being called from `StoreRecovery` class which is in a different package.

File: server/src/main/java/org/opensearch/index/translog/TranslogHeader.java [hunk 4]
```diff
@@ -137,9 +137,26 @@ static int readHeaderVersion(final Path path, final FileChannel channel, final S
...
+        // verify UUID only after checksum, to ensure that UUID is not corrupted
+        final BytesRef expectedUUID = new BytesRef(translogUUID);
+        final BytesRef actualUUID = new BytesRef(translogHeader.translogUUID);
+        if (actualUUID.bytesEquals(expectedUUID) == false) {
+            throw new TranslogCorruptedException(
```
[the remark this answers is not in the document]
Comment (sachinpkale) [reply server/src/main/java/org/opensearch/index/translog/TranslogHeader.java:148]: Assuming you are referring to following comment:
```
// verify UUID only after checksum, to ensure that UUID is not corrupted
```
It needs to be for this method as `translogUUID` is one of the arguments. For the other overload of `read`, we don't have a UUID to compare against.

File: server/src/main/java/org/opensearch/index/translog/TranslogHeader.java [hunk 5]
```diff
@@ -137,9 +137,26 @@ static int readHeaderVersion(final Path path, final FileChannel channel, final S
...
+                "expected shard UUID " + expectedUUID + " but got: " + actualUUID + " this translog file belongs to a different translog"
+            );
+        }
+        return translogHeader;
+    }
+
+    /**
+     * Read a translog header from the given path and file channel and compare the given UUID
+     */
+    public static TranslogHeader read(final Path path, final FileChannel channel) throws IOException {
```
[the remark this answers is not in the document]
Comment (sachinpkale) [reply server/src/main/java/org/opensearch/index/translog/TranslogHeader.java:159]: Yeah, needs to be public as being called from `StoreRecovery` class which is in a different package.

File: server/src/main/java/org/opensearch/index/translog/TranslogHeader.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (BhumikaSaini-Amazon) [reply server/src/main/java/org/opensearch/index/translog/TranslogHeader.java:148]: Yes, that is the comment. Okay. I saw the checksum validation logic in the other overload followed by invocation of this method and hence proposed this. We can resolve this thread.

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreRestoreIT.java [hunk 6]
```diff
@@ -186,6 +184,8 @@ private void testRestoreFlow(int numberOfIterations, boolean invokeFlush, int sh
         Map<String, Long> indexStats = indexData(numberOfIterations, invokeFlush, INDEX_NAME);
         assertEquals(shardCount, getNumShards(INDEX_NAME).totalNumShards);
 
+        assertHitCount(client().prepareSearch(INDEX_NAME).setSize(0).get(), indexStats.get(REFRESHED_OR_FLUSHED_OPERATIONS));
```
[the remark this answers is not in the document]
Comment (BhumikaSaini-Amazon) [reply server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreRestoreIT.java:187]: Got it. I was comparing this with what we were doing in `verifyRestoredData()`. But I see that we have an explicit refresh there after the restore. So this looks correct, thanks!

**Comment by sachinpkale** [comment]
The following tests are showing flaky behavior:

```
Tests with failures:
 - org.opensearch.indices.replication.SegmentReplicationSuiteIT.testFullRestartDuringReplication
 - org.opensearch.indices.replication.SegmentReplicationSuiteIT.testDropRandomNodeDuringReplication
 - org.opensearch.indices.replication.SegmentReplicationSuiteIT.testBasicReplication
 - org.opensearch.indices.replication.SegmentReplicationSuiteIT.classMethod
```

Checking if the flakiness is introduced with this PR.

**Comment by sachinpkale** [comment]
> The following tests are showing flaky behavior:
> 
> ```
> Tests with failures:
>  - org.opensearch.indices.replication.SegmentReplicationSuiteIT.testFullRestartDuringReplication
>  - org.opensearch.indices.replication.SegmentReplicationSuiteIT.testDropRandomNodeDuringReplication
>  - org.opensearch.indices.replication.SegmentReplicationSuiteIT.testBasicReplication
>  - org.opensearch.indices.replication.SegmentReplicationSuiteIT.classMethod
> ```
> 
> Checking if the flakiness is introduced with this PR.

These are flaky tests and able to reproduce the failure on main as well. Issue: https://github.com/opensearch-project/OpenSearch/issues/9499

**Comment by sachinpkale** [comment]
Added unit tests. `StoreRecovery` changes are covered in integ tests.

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreRestoreIT.java
Comment (BhumikaSaini-Amazon) [inline server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreRestoreIT.java:1]: nit : 

Please could you help remove the `awaitsFix` on 6188 from `testRTSRestoreWithNoDataPostCommitPrimaryReplicaDown` as this PR is resolving that issue? 
We can cover this as part of #8951

File: server/src/test/java/org/opensearch/index/store/StoreTests.java [hunk 7]
```diff
@@ -1166,6 +1167,42 @@ public void testGetMetadataWithSegmentInfos() throws IOException {
...
+        Store store = new Store(shardId, INDEX_SETTINGS, new NIOFSDirectory(createTempDir()), new DummyShardLock(shardId));
+        store.createEmpty(Version.LATEST, null);
+        SegmentInfos segmentInfos = Lucene.readSegmentInfos(store.directory());
+        assertFalse(segmentInfos.getUserData().containsKey(Translog.TRANSLOG_UUID_KEY));
+        assertEquals("-1", segmentInfos.getUserData().get(SequenceNumbers.LOCAL_CHECKPOINT_KEY));
+        assertEquals("-1", segmentInfos.getUserData().get(SequenceNumbers.MAX_SEQ_NO));
+        assertEquals("-1", segmentInfos.getUserData().get(Engine.MAX_UNSAFE_AUTO_ID_TIMESTAMP_COMMIT_ID));
+        store.close();
+    }
+
```
Comment (BhumikaSaini-Amazon) [inline server/src/test/java/org/opensearch/index/store/StoreTests.java:1205]: nit :

Should we extract the common assertions into a helper?

**Review by BhumikaSaini-Amazon — APPROVED** [review]

File: server/src/main/java/org/opensearch/index/translog/TranslogHeader.java (hunk 5 above)
Comment (gbbafna) [inline server/src/main/java/org/opensearch/index/translog/TranslogHeader.java:159]: lets not make this public as it is not meant to be consumed

Comment (gbbafna) [inline server/src/main/java/org/opensearch/index/translog/TranslogHeader.java:159]:
```suggestion
    static TranslogHeader read(final Path path, final FileChannel channel) throws IOException {
```

Comment (sachinpkale) [reply server/src/main/java/org/opensearch/index/translog/TranslogHeader.java:159]: This needs to be public as `StoreRecovery` is calling with the change in this PR.

**Review by gbbafna — APPROVED** [review]

**Pull request merged**